### PR TITLE
feat(inertia): migrate video stream page to Inertia

### PR DESCRIPTION
**Pull request by mxrcabrera**
Issue: #3141

# Description

Migrates the video stream page (`url_redirects#stream`) from ReactOnRails server component rendering to Inertia.js.

- Created new Inertia page component `UrlRedirects/Stream.tsx`
- Updated `UrlRedirectsController#stream` to render via Inertia with props instead of ERB template
- Removed old `video_stream.html.erb` view and `video_stream.ts` pack
- Removed server component registration from `VideoStreamPlayer` (now used as regular React component)
- Updated controller specs to use Inertia test helpers
- **Added custom domain support test as explicitly required by issue**

---

# Before/After

## Before

https://github.com/user-attachments/assets/0498c0c0-f727-47ea-ae88-6db57d5dd9e5

## After

After app url

https://github.com/user-attachments/assets/675fbba6-b46b-4541-bfc9-7a034e04d447

After custom url

https://github.com/user-attachments/assets/31cbc929-e09c-4a62-91bf-46470435de83

---

# Test Results

image

---

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/antiwork/gumroad/blob/main/CONTRIBUTING.md)
- [x] I have watched [Gumroad PR review livestreams](https://www.youtube.com/@anti-work)
- [x] I have performed a self-review and left review comments on my PR
- [x] I have added/updated tests for my changes

---

# AI Disclosure

IDE: VS Code with Claude Code  
Model: Claude Sonnet 4  
Purpose: Migration implementation and testing

### CODE REVIEW

File: app/controllers/url_redirects_controller.rb [hunk 1]
```diff
@@ -235,18 +235,20 @@ def send_to_kindle
...
+    e404 unless product_file&.streamable?
 
-    @videos_playlist = @url_redirect.video_files_playlist(@product_file)
-    @should_show_transcoding_notice = logged_in_user == @url_redirect.seller && !@url_redirect.with_product_files.has_been_transcoded?
+    videos_playlist = @url_redirect.video_files_playlist(product_file)
 
-    @url_redirect_id = @url_redirect.external_id
-    @purchase_id = @url_redirect.purchase.try(:external_id)
-    render :video_stream
+    render inertia: "UrlRedirects/Stream", props: {
```
Comment (mxrcabrera) [inline app/controllers/url_redirects_controller.rb]: Replaced ERB render with Inertia, passing all props needed by VideoStreamPlayer

File: app/javascript/pages/UrlRedirects/Stream.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        url_redirect_id={url_redirect_id}
+        purchase_id={purchase_id}
+        should_show_transcoding_notice={should_show_transcoding_notice}
+        transcode_on_first_sale={transcode_on_first_sale}
+      />
+    </div>
+  );
+}
+
+StreamPage.loggedInUserLayout = true;
```
Comment (mxrcabrera) [inline app/javascript/pages/UrlRedirects/Stream.tsx:167]: Using loggedInUserLayout to maintain user context

File: spec/controllers/url_redirects_controller_spec.rb [hunk 3]
```diff
@@ -583,6 +584,22 @@
         end
       end
 
+      context "with custom domain" do
```
Comment (mxrcabrera) [inline spec/controllers/url_redirects_controller_spec.rb:593]: Added as explicitly required by issue #3141

**Comment by Pradumn27** [comment]
@mxrcabrera _a, can you please resolve the conflicts?

**Comment by mxrcabrera** [comment]
@Pradumn27 Done, conflicts resolved!

**Comment by mxrcabrera** [comment]
@Pradumn27 could you please review this PR? Thanks

**Comment by EmCousin** [comment]
Please fix the conflicts

**Comment by mxrcabrera** [comment]
@EmCousin Done, conflicts resolved!

File: app/controllers/url_redirects_controller.rb [hunk 4]
```diff
@@ -237,18 +237,20 @@ def send_to_kindle
...
+      playlist: videos_playlist[:playlist],
+      index_to_play: videos_playlist[:index_to_play].to_i,
+      url_redirect_id: @url_redirect.external_id,
+      purchase_id: @url_redirect.purchase.try(:external_id),
+      should_show_transcoding_notice: logged_in_user == @url_redirect.seller && !@url_redirect.with_product_files.has_been_transcoded?,
+      transcode_on_first_sale: product_file.link&.transcode_videos_on_purchase.present?,
+    }
```
Comment (Pradumn27) [inline app/controllers/url_redirects_controller.rb]: Create a custom function called `steam_page_props` in the `url_redirect_presenter` and do the prop processing logic there, and use that presenter here

File: app/javascript/pages/UrlRedirects/Stream.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  return (
+    <div id="stream_page" className="download-page responsive responsive-nav absolute inset-0">
+      <VideoStreamPlayer
+        playlist={playlist}
+        index_to_play={index_to_play}
+        url_redirect_id={url_redirect_id}
+        purchase_id={purchase_id}
+        should_show_transcoding_notice={should_show_transcoding_notice}
+        transcode_on_first_sale={transcode_on_first_sale}
+      />
```
Comment (Pradumn27) [inline app/javascript/pages/UrlRedirects/Stream.tsx]: directly inline the component here, won't need this extra wrapper layer

File: spec/controllers/url_redirects_controller_spec.rb [hunk 6]
```diff
@@ -563,11 +563,12 @@
...
         end
 
-        it "assigns the first streamable product file to '@product_file' and renders the stream page correctly" do
+        it "renders the Inertia stream page with playlist props" do
           get :stream, params: { id: @url_redirect.token }
 
           expect(response).to have_http_status(:ok)
-          expect(assigns(:product_file)).to eq(@product.product_files.first)
+          expect(inertia.component).to eq("UrlRedirects/Stream")
+          expect(inertia.props[:playlist]).to be_present
```
Comment (Pradumn27) [inline spec/controllers/url_redirects_controller_spec.rb:572]: let's check for all the props and their values

File: spec/controllers/url_redirects_controller_spec.rb [hunk 7]
```diff
@@ -997,27 +1014,27 @@
...
 
           expect(response).to be_successful
-          expect(assigns(:hide_layouts)).to eq(true)
+          expect(inertia.component).to eq("UrlRedirects/Stream")
+          video_urls = inertia.props[:playlist]
           expect(video_urls.size).to eq 1
           expect(video_urls[0][:sources][0]).to include "index.m3u8"
           expect(video_urls[0][:sources][1]).to include @video_file_1.s3_filename
-          expect(assigns(:videos_playlist)[:index_to_play]).to eq 0
+          expect(inertia.props[:index_to_play]).to eq 0
```
Comment (Pradumn27) [inline spec/controllers/url_redirects_controller_spec.rb:1031]: test for all the props

Comment (mxrcabrera) [reply spec/controllers/url_redirects_controller_spec.rb:1031]: image

**Comment by mxrcabrera** [comment]
@Pradumn27 Done, all feedback addressed.

**Comment by 4arjun** [comment]
@mxrcabrera This branch has multiple ESLint errors. Could you please run ESLint locally and fix them?

**Comment by mxrcabrera** [comment]
@4arjun Fixed in dd57dbb - rebased on latest main and ran ESLint/TypeScript/Rubocop locally, all passing now.

File: app/controllers/url_redirects_controller.rb [hunk 8]
```diff
@@ -237,18 +237,12 @@ def send_to_kindle
...
 
-    @videos_playlist = @url_redirect.video_files_playlist(@product_file)
-    @should_show_transcoding_notice = logged_in_user == @url_redirect.seller && !@url_redirect.with_product_files.has_been_transcoded?
+    product_file = @url_redirect.product_file(params[:product_file_id]) || @url_redirect.alive_product_files.find(&:streamable?)
+    e404 unless product_file&.streamable?
 
-    @url_redirect_id = @url_redirect.external_id
-    @purchase_id = @url_redirect.purchase.try(:external_id)
-    render :video_stream
+    props = UrlRedirectPresenter.new(url_redirect: @url_redirect, logged_in_user:).stream_page_props(product_file:)
```
Comment (4arjun) [inline app/controllers/url_redirects_controller.rb:244]: Let’s inline this below?

File: app/javascript/pages/UrlRedirects/Stream.tsx [hunk 9]
```diff
@@ -29,23 +29,27 @@ type Video = {
...
   playlist: Video[];
   index_to_play: number;
   url_redirect_id: string;
   purchase_id: string | null;
   should_show_transcoding_notice: boolean;
   transcode_on_first_sale: boolean;
-}) => {
+};
+
+const fakeVideoUrlGuidForObfuscation = "ef64f2fef0d6c776a337050020423fc0";
```
Comment (4arjun) [inline app/javascript/pages/UrlRedirects/Stream.tsx]: Better to declare constants at the top of the file and in capital snake case -`FAKE_VIDEO_URL_GUID_FOR_OBFUSCATION`

File: app/controllers/url_redirects_controller.rb (hunk 8 above)
Comment (mxrcabrera) [reply app/controllers/url_redirects_controller.rb:244]: The presenter is shared across 7 controller actions (stream, read, download, etc.) - it's not video-stream-specific. Inlining would bloat the controller and break the existing presenter pattern used across the project.

**Comment by EmCousin** [comment]
Solved in https://github.com/antiwork/gumroad/pull/3382

**Comment by mxrcabrera** [comment]
@EmCousin I'm genuinely disappointed by how this was handled.

Looking at the timeline: I was second to submit (after #3156), my PR #3241 came 2 weeks ago, 4arjun's #3382 came last week.

For two weeks I responded to every review within hours, resolved conflicts multiple times, and addressed all requested changes. When @Pradumn27 asked about the presenter pattern, I implemented `stream_page_props` and explained my reasoning.

The closing comment says #3382 had "the cleanest code changes and the most thorough test coverage." But comparing the two PRs:

- My PR: 7 files changed, 3 commits
- #3382: 8 files changed, 21 commits

Both implement `stream_page_props` in the presenter. Both test all props. Both include custom domain tests. My PR was actually cleaner by the numbers.

I'm not claiming my implementation was perfect. But being closed without any feedback about what was actually wrong - while a PR with more commits and more files gets merged as "cleanest" - is confusing.

What could I have done differently here?

**Comment by EmCousin** [comment]
@mxrcabrera The deciding factor was test coverage. #3382 included a dedicated presenter spec testing stream_page_props in isolation, covering cases like transcoding notice when the seller is logged in, purchase_id when a purchase is present, transcode_on_first_sale enabled, and multiple video files, plus an edge case test for non-streamable files returning 404. Your PR tested the default values but didn't cover these alternate states.
